pkg/controllermanager/controller/controllerregistration/controllerregistration_control.go

```diff
@@ -0,0 +1,272 @@
+	}
+
+	c.controllerRegistrationQueue.AddAfter(key, 30*time.Second)
+	return nil
+}
+
+// ControlInterface implements the control logic for updating ControllerRegistrations. It is implemented as an interface to allow
+// for extensions that provide different semantics. Currently, there is only one implementation.
+type ControlInterface interface {
+	Reconcile(*corev1alpha1.ControllerRegistration) error
+}
+
+// NewDefaultControllerRegistrationControl returns a new instance of the default implementation ControlInterface that
+// implements the documented semantics for ControllerRegistrations. updater is the UpdaterInterface used
+// to update the status of ControllerRegistrations. You should use an instance returned from NewDefaultControllerRegistrationControl() for any
+// scenario other than testing.
+func NewDefaultControllerRegistrationControl(k8sGardenClient kubernetes.Interface, k8sGardenInformers gardeninformers.SharedInformerFactory, k8sGardenCoreInformers gardencoreinformers.SharedInformerFactory, recorder record.EventRecorder, config *config.ControllerManagerConfiguration, seedLister gardenlisters.SeedLister, controllerRegistrationLister gardencorelisters.ControllerRegistrationLister, controllerInstallationLister gardencorelisters.ControllerInstallationLister) ControlInterface {
+	return &defaultControllerRegistrationControl{k8sGardenClient, k8sGardenInformers, k8sGardenCoreInformers, recorder, config, seedLister, controllerRegistrationLister, controllerInstallationLister}
+}
+
+type defaultControllerRegistrationControl struct {
+	k8sGardenClient              kubernetes.Interface
+	k8sGardenInformers           gardeninformers.SharedInformerFactory
+	k8sGardenCoreInformers       gardencoreinformers.SharedInformerFactory
+	recorder                     record.EventRecorder
+	config                       *config.ControllerManagerConfiguration
+	seedLister                   gardenlisters.SeedLister
+	controllerRegistrationLister gardencorelisters.ControllerRegistrationLister
+	controllerInstallationLister gardencorelisters.ControllerInstallationLister
+}
+
+func (c *defaultControllerRegistrationControl) Reconcile(obj *corev1alpha1.ControllerRegistration) error {
+	var (
+		controllerRegistration = obj.DeepCopy()
+		logger                 = logger.NewFieldLogger(logger.Logger, "controllerregistration", controllerRegistration.Name)
+	)
+
+	logger.Infof("[CONTROLLERREGISTRATION RECONCILE] %s", controllerRegistration.Name)
+
+	if controllerRegistration.DeletionTimestamp != nil {
+		return c.delete(controllerRegistration, logger)
+	}
+
+	return c.reconcile(controllerRegistration, logger)
+}
+
+func (c *defaultControllerRegistrationControl) reconcile(controllerRegistration *corev1alpha1.ControllerRegistration, logger logrus.FieldLogger) error {
+	var (
+		err              error
+		result           error
+		installationsMap = map[string]string{}
+
+		mustWriteFinalizer = false
+	)
+
+	seedList, err := c.seedLister.List(labels.Everything())
```

Comment:
Installing on each seed seems a bit tough - can't we already have a `selector` of some kind to limit this?

Reply:
We decided to go with installation on all seeds for now and implement a smarter logic with a separate PR (need discussion about details of this "smart logic").